resolver/dns_resolver_cache.ml

```diff
@@ -11,6 +11,47 @@ let _pp_err ppf = function
 let pp_question ppf (name, typ) =
   Fmt.pf ppf "%a (%a)" Domain_name.pp name Packet.Question.pp_qtype typ
 
+let nsec_no_ds t ts name =
+  let rec up name =
+    match snd (Dns_cache.get t ts name Nsec) with
+    | Ok (`Entry (_, nsec), _) ->
+      Bit_map.mem (Rr_map.to_int Ds) nsec.Nsec.types
+    | _ ->
+      if Domain_name.count_labels name >= 1 then
+        up (Domain_name.drop_label_exn name)
+      else
+        false
+  in
+  up name
+
+let nsec3_covering t ts name =
+  let rec up name =
+    match snd (Dns_cache.get_nsec3 t ts name) with
+    | Ok nsec3 ->
+      let Nsec3.{ iterations ; salt ; _ } = snd (List.hd nsec3) in
+      let soa_name = Domain_name.drop_label_exn ~rev:true name in
```

Comment:
Do we really want `~rev:true`? doesn't that turn foo.example.org into foo.example?

Reply:
you're right, I'm not sure what I had in mind here.